src/SourceFile.js

```diff
@@ -36,31 +37,59 @@ import NotImplementedError from "./NotImplementedError.js";
  */
 class SourceFile {
     /**
-     * Construct a new source file instance. 
+     * Construct a new source file instance.
      *
      * @param {String} uri URI or path to the source file
      * @param {Object} [options] options to the constructor
+     * @param {SourceFile} [options.file] the source file to copy from. Other options
+     * will override the fields in this file
      * @param {String} [options.sourceLocale] the source locale of the files
      * being linted
      * @param {Function} [options.getLogger] a callback function provided by
      * the linter to retrieve the log4js logger
-     * @param [String] [options.type] the type of this file
+     * @param {String} [options.type] the type of this file
+     * @param {String} [options.content] the content of the file
      * @constructor
      */
     constructor(uri, options) {
         if (!uri) {
             throw new Error("Attempt to create a SourceFile without a file path");
         }
         this.filePath = uri;
-        this.logger = (typeof(options.getLogger) === "function") ? options.getLogger("ilib-lint-common.SourceFile") : undefined;
-        this.sourceLocale = options?.sourceLocale || "en-US";
-        this.type = options?.type || "";
+        if (options?.file) {
+            this.logger = options.file.logger;
+            this.sourceLocale = options.file.sourceLocale;
+            this.raw = options.file.raw;
+            this.content = options.file.content;
+            this.dirty = options.file.dirty;
+        }
```

Comment:
Could this be factored out into a static method like `SourceFile.copy()`?

Reply:
You don't like copy constructors? Too Java-ish? ;-)

Reply:
Nah it would be fine if we could have overloaded constructors like
```js
constructor(file: SourceFile) {}
constructor(options: Options) {}
```
but I don't like mixing two different ways of setup in the same function